erpnext/hr/doctype/salary_slip/salary_slip.py

```diff
@@ -20,53 +20,103 @@ def autoname(self):
+
 			if struct:
-				self.set("earnings", [])
-				self.set("deduction", [])
-				self.pull_sal_struct(struct)
+				ss_doc = frappe.get_doc('Salary Structure', struct)
+				self.salary_slip_based_on_timesheet = ss_doc.salary_slip_based_on_timesheet or 0
+				self.set_time_sheet()
+				self.pull_sal_struct(ss_doc)
+
+	def set_time_sheet(self):
+		if self.salary_slip_based_on_timesheet and not self.get('timesheets'):
+			self.set("timesheets", [])
+
+			timesheets = frappe.db.sql(""" select * from `tabTime Sheet` where employee = %(employee)s and (status = 'Submitted' or
+				status = 'Billed') and %(start_date)s between start_date and end_date and %(end_date)s between start_date and end_date""", 
+				{'start_date': self.start_date, 'end_date': self.end_date, 'employee': self.employee}, as_dict=1)
+				
+			for data in timesheets:
+				self.append('timesheets', {
+					'time_sheet': data.name,
+					'working_hours': data.total_hours
+				})
 
-	def check_sal_struct(self, joining_date, relieving_date):
-		m = get_month_details(self.fiscal_year, self.month)
+	def set_month_dates(self):
+		if self.month and not self.salary_slip_based_on_timesheet:
+			m = get_month_details(self.fiscal_year, self.month)
+			self.start_date = m['month_start_date']
+			self.end_date = m['month_end_date']
 
+	def check_sal_struct(self, joining_date, relieving_date):
+		timesheet = 1 if self.salary_slip_based_on_timesheet else 0
 		struct = frappe.db.sql("""select name from `tabSalary Structure`
 			where employee=%s and is_active = 'Yes'
 			and (from_date <= %s or from_date <= %s)
-			and (to_date is null or to_date >= %s or to_date >= %s)""",
-			(self.employee, m.month_start_date, joining_date, m.month_end_date, relieving_date))
+			and (to_date is null or to_date >= %s or to_date >= %s) and salary_slip_based_on_timesheet=%s""",
+			(self.employee, self.start_date, joining_date, self.end_date, relieving_date, timesheet))
 
 		if not struct:
-			msgprint(_("No active Salary Structure found for employee {0} and the month")
+			self.salary_structure = None
+			msgprint(_("No active or default Salary Structure found for employee {0} and the month")
 				.format(self.employee))
-			self.employee = None
 
 		return struct and struct[0][0] or ''
 
-	def pull_sal_struct(self, struct):
+	def pull_sal_struct(self, ss_doc):
 		from erpnext.hr.doctype.salary_structure.salary_structure import make_salary_slip
-		make_salary_slip(struct, self)
+		make_salary_slip(ss_doc.name, self)
+
+		if self.salary_slip_based_on_timesheet:
+			self.salary_structure = ss_doc.name
+			self.hour_rate = ss_doc.hour_rate
+			self.total_working_hours = sum([d.working_hours or 0.0 for d in self.timesheets]) or 0.0
+			self.add_earning_for_hourly_wages(ss_doc.earning_type)
+
+	def add_earning_for_hourly_wages(self, earning_type):
+		default_type = False
+		for data in self.earnings:
+			if data.e_type == earning_type:
+				data.e_modified_amount = self.hour_rate * self.total_working_hours
```

Comment:
Can I change e_modified_amount to earning_amount and e_type to earning_type
d_type to deduction_type and d_modified_amount to deduction_amount